test/jdk/java/text/Format/MessageFormat/MessageFormatToPatternTest.java

```diff
@@ -0,0 +1,323 @@
+        testRoundTrip(new MessageFormat("Test: {0,number,foo'{'#.00}"));
+        testRoundTrip(new MessageFormat("Test: {0,number,foo'}'#.00}"));
+    }
+
+    // Go roundrip from MessageFormat -> pattern string -> MessageFormat and verify equivalence
+    @ParameterizedTest
+    @MethodSource("testCases")
+    public void testRoundTrip(MessageFormat format1) {
+
+        // Prepare MessageFormat argument
+        Object[] args = new Object[] {
+            8.5,                            // argument for DecimalFormat
+            new Date(1705502102677L),       // argument for SimpleDateFormat
+            random.nextInt(6)               // argument for ChoiceFormat
+        };
+
+        String pattern1 = null;
+        String result1 = null;
+        String pattern2 = null;
+        String result2 = null;
+        try {
+
+            // Format using the given MessageFormat
+            pattern1 = format1.toPattern();
+            result1 = format1.format(args);
+
+            // Round-trip via toPattern() and repeat
+            MessageFormat format2 = new MessageFormat(pattern1);
+            pattern2 = format2.toPattern();
+            result2 = format2.format(args);
+
+            // Check equivalence
+            assertEquals(result1, result2);
+            assertEquals(pattern1, pattern2);
+        } catch (RuntimeException | Error e) {
+            System.out.println(String.format("%n********** FAILURE **********%n"));
+            System.out.println(String.format("%s%n", e));
+            if (!spitSeed) {
+                System.out.println(String.format("*** Random seed was 0x%016xL%n", randomSeed));
+                spitSeed = true;
+            }
+            print(0, format1);
+            System.out.println();
+            if (pattern1 != null)
+                System.out.println(String.format("  pattern1 = %s", javaLiteral(pattern1)));
+            if (result1 != null)
+                System.out.println(String.format("   result1 = %s", javaLiteral(result1)));
+            if (pattern2 != null)
+                System.out.println(String.format("  pattern2 = %s", javaLiteral(pattern2)));
+            if (result2 != null)
+                System.out.println(String.format("   result2 = %s", javaLiteral(result2)));
+            System.out.println();
+            throw e;
+        }
+    }
+
+    public static Stream<Arguments> testCases() {
+        final ArrayList<Arguments> argList = new ArrayList<>();
+        for (int i = 0; i < NUM_RANDOM_TEST_CASES; i++)
+            argList.add(Arguments.of(randomFormat()));
+        return argList.stream();
+    }
+
+    // Generate a "random" MessageFormat. We do this by creating a MessageFormat with "{0}" placeholders
+    // and then substituting in random DecimalFormat, DateFormat, and ChoiceFormat subformats. The goal here
+    // is to avoid using pattern strings to construct formats, because they're what we're trying to check.
```

Comment:
Can you add a general example of a randomly generated MessageFormat in the comments, (I know they vary by quite a lot), but it would be hard for someone to piece if together without digging into the code. And unfortunately the current toString() of MessageFormat is not implemented, so although JUnit prints the MessageFormat out, it's just gibberish.

Reply:
Added in 58e8cc68f45 as a test case.